File: old/workspace-old/w13_photo_proj_2022/src/main/scala/ImageFilter.scala
package photo

import introprog.Image
/**
 * (2021-07-25) translate from java. Replaced nbrOfArgs with args.
 * (Theodor Lundqvist)
 * 
 * (2016-07-17) nbrOfArgs attribut has been added.
 * (Casper Schreiter, Björn Regnell)
 * 
 */

/**
 * The super class to all filters.
 * Create a filter object with a given name and argument descriptions.
 * @param name
 *            the name of the filter.
 * @param args
 *            optional array of strings with argument descriptions
 */
abstract class ImageFilter(val name: String, val argDescriptions: String*):

    /**The number of args this filter needs*/
    def nbrOfArgs = argDescriptions.length

    /**
     * Apply the filter on `img` and return the result as a new Image using the arguments in `args`.
        * 
        * @param img
        *            the original image.
        * @param args
        *            arguments
        * @return the resulting image.
        */
    def apply(img: Image, args: Double*): Image;

    /**
     * Calculate the intensity in each pixel of `img`.
        * 
        * @param img
        *           the image
        * @return intensitymatrix, values ranging from 0 to 255
        */
    protected def computeIntensity(img: Image): Array[Array[Short]] = 
        val intensity : Array[Array[Short]] = Array.ofDim(img.width, img.height)
        for 
            w <- 0 until img.width
            h <- 0 until img.height 
        do
            val c = img(w, h)
            intensity(w)(h) = ((c.getRed()+c.getGreen+c.getBlue())/3).toShort
        intensity

    /**
     * Convolute `p[i][j]` with the convolutionkernel `kernel`.
        * 
        * @param p
        *            matrix with numbervalues
        * @param i
        *            current row index
        * @param j
        *            current coloumn index
        * @param kernel
        *            convolutionkernel, a 3x3-matrix
        * @param weight
        *            the sum of the element in `kernel`
        * @return result of the convolution
        */
    protected def convolve(p: Array[Array[Short]], i: Int, j: Int, kernel: Array[Array[Short]],
            weight: Int): Short =
        var sum : Double = 0;

        for ii <- -1 to 1 do
            for jj <- -1 to 1 do
                sum += p(i + ii)(j + jj) * kernel(ii + 1)(jj + 1);

        Math.round(sum / weight).toShort;

File: workspace-en/cslib/src/main/java/cslib/images/ImageFilter.java
package cslib.images;

import java.awt.Color;

/**
 * Superclass for all filter classes.
 * @version 1.2 (2016-07-17) nbrOfArgs attribute has been added 
 * along with a getNumberOfArguments-method (Casper Schreiter, Björn Regnell)
 */
public abstract class ImageFilter {
	private String name; 		// filtrets namn
	private int nbrOfArgs; 		// number of arguments

	/**
	 * Creates a filter object with a given name and the number of arguments the filter requires.
	 * 
	 * @param name
	 *            the filter's name
	 * @param nbrOfArgs
	 *            the number of arguments
	 */
	protected ImageFilter(String name, int nbrOfArgs) {
		this.name = name;
		this.nbrOfArgs = nbrOfArgs;		
	}
	

	/**
	 * Tar reda på filtrets namn.
	 * 
	 * @return filtrets namn
	 */
	public String getName() {
		return name;
	}

	/**
 * Determines the number of arguments the filter requires.
 * 
 * @return the number of arguments needed by the apply method.
	 */
	public int getNumberOfArguments() {
		return nbrOfArgs;
	}

	/**
	 * Filtering the image in the matrix inPixels and returning the result in a new
	 * matrix. May utilize values from args
	 * 
	 * @param inPixels
	 *            the original image
	 * @param args
	 *            arguments
	 * @return the filtered image
	 */
	public abstract Color[][] apply(Color[][] inPixels, double[] args);
	
	/**
	 * Calculates the intensity of all pixels in pixels, returning the result in
	 * a new matrix.
	 * 
	 * @param pixels
	 *            matrix with pixels
	 * @return the intensity of each pixel

*
	 */
	protected short[][] computeIntensity(Color[][] pixels) {
		int height = pixels.length;
		int width = pixels[0].length;
		short[][] intensity = new short[height][width];
		for (int i = 0; i < height; i++) {
			for (int j = 0; j < width; j++) {
				Color c = pixels[i][j];
				intensity[i][j] = (short) ((c.getRed() + c.getGreen() + c
						.getBlue()) / 3);
			}
		}
		return intensity;
	}

	/**
	 * Missing the point p[i][j] with folding key kernel.
	 * 
	 * @param p
	 *            matrix with numerical values
	 * @param i
	 *            row index for the current point
	 * @param j
	 *            column index for the current point
	 * @param kernel
	 *            the folding key, a 3x3-matrix
	 * @param weight
	 *            sum of elements in kernel
	 * @return result of the convolution
	 */
	protected short convolve(short[][] p, int i, int j, short[][] kernel,
			int weight) {
		short sum = 0;
		for (int ii = -1; ii <= 1; ii++) {
			for (int jj = -1; jj <= 1; jj++) {
				sum += p[i + ii][j + jj] * kernel[ii + 1][jj + 1];
			}
		}
		return (short) Math.round((double) sum / weight);
	}
}